File: OpenSpiel/oware/oware.cc
// Copyright 2019 DeepMind Technologies Limited
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "open_spiel/games/oware/oware.h"

#include <iomanip>

#include "open_spiel/game_parameters.h"

namespace open_spiel {
namespace oware {

namespace {

// Facts about the game as required by OpenSpiel. In particular, Oware is a
// deterministic, sequential, two-player, perfect-information, zero-sum game.
const GameType kGameType{
    /*short_name=*/"oware",
    /*long_name=*/"Oware",
    GameType::Dynamics::kSequential,
    GameType::ChanceMode::kDeterministic,
    GameType::Information::kPerfectInformation,
    GameType::Utility::kZeroSum,
    GameType::RewardModel::kTerminal,
    /*max_num_players=*/2,
    /*min_num_players=*/2,
    /*provides_information_state_string=*/false,
    /*provides_information_state_tensor=*/false,
    /*provides_observation_string=*/true,
    /*provides_observation_tensor=*/true,
    /*parameter_specification=*/
    {{"num_houses_per_player", GameParameter(kDefaultHousesPerPlayer)},
     {"num_seeds_per_house", GameParameter(kDdefaultSeedsPerHouse)}}};

std::shared_ptr<const Game> Factory(const GameParameters& params) {
  return std::shared_ptr<const Game>(new OwareGame(params));
}

REGISTER_SPIEL_GAME(kGameType, Factory);

RegisterSingleTensorObserver single_tensor(kGameType.short_name);

}  // namespace

OwareState::OwareState(std::shared_ptr<const Game> game,
                       int num_houses_per_player, int num_seeds_per_house)
    : State(game),
      num_houses_per_player_(num_houses_per_player),
      total_seeds_(kNumPlayers * num_seeds_per_house * num_houses_per_player),
      board_(/*num_houses_per_player=*/num_houses_per_player,
             /*num_seeds_per_house=*/num_seeds_per_house) {
  // Repetition is tracked from the beginning, including the initial board.
  boards_since_last_capture_.insert(board_);
}

OwareState::OwareState(std::shared_ptr<const Game> game,
                       const OwareBoard& board)
    : State(game),
      num_houses_per_player_(board.seeds.size() / kNumPlayers),
      total_seeds_(board.TotalSeeds()),
      board_(board) {
  SPIEL_CHECK_EQ(0, board.seeds.size() % kNumPlayers);
  SPIEL_CHECK_TRUE(IsTerminal() || !LegalActions().empty());
  boards_since_last_capture_.insert(board_);
}

std::vector<Action> OwareState::LegalActions() const {
  std::vector<Action> actions;
  if (IsTerminal()) return actions;
  // Translate the current player's row into physical indices in seeds[].
  const Player lower = PlayerLowerHouse(board_.current_player);
  const Player upper = PlayerUpperHouse(board_.current_player);
  if (OpponentSeeds() == 0) {
    // In case the opponent does not have any seeds, a player must make
    // a move which gives the opponent seeds.
    // If the opponent is empty, a move is legal only when sowing reaches the
    // opponent's row and gives it at least one seed.
    for (int house = lower; house <= upper; house++) {
      const int first_seeds_in_own_row = upper - house;
      if (board_.seeds[house] - first_seeds_in_own_row > 0) {
        actions.push_back(HouseToAction(house));
      }
    }
  } else {
    // In the normal case, every non-empty house in the current player's row
    // is a legal source for sowing.
    for (int house = lower; house <= upper; house++) {
      if (board_.seeds[house] > 0) {
        actions.push_back(HouseToAction(house));
      }
    }
  }
  return actions;
}

std::string OwareState::ActionToString(Player player, Action action) const {
  // Player 0's actions are displayed as A-F and player 1's as a-f.
  return std::string(1, (player == Player{0} ? 'A' : 'a') + action);
}

void OwareState::WritePlayerScore(std::ostringstream& out,
                                  Player player) const {
  out << "Player " << player << " score = " << board_.score[player];
  if (CurrentPlayer() == player) {
    out << " [PLAYING]" << std::endl;
  } else {
    out << std::endl;
  }
}

std::string OwareState::ToString() const {
  std::ostringstream out;
  if (IsTerminal()) {
    out << "[FINISHED]" << std::endl;
  }
  WritePlayerScore(out, 1);

  // Add player 1 labels.
  for (int action = num_houses_per_player_ - 1; action >= 0; action--) {
    out << std::setw(3) << std::right << ActionToString(1, action);
  }
  out << std::endl;

  // Add player 1 house seeds.
  for (int house = kNumPlayers * num_houses_per_player_ - 1;
       house >= num_houses_per_player_; house--) {
    out << std::setw(3) << std::right << board_.seeds[house];
  }
  out << std::endl;

  // Add player 0 house seeds.
  for (int house = 0; house < num_houses_per_player_; house++) {
    out << std::setw(3) << std::right << board_.seeds[house];
  }
  out << std::endl;

  // Add player 0 labels.
  for (int action = 0; action < num_houses_per_player_; action++) {
    out << std::setw(3) << std::right << ActionToString(0, action);
  }
  out << std::endl;

  WritePlayerScore(out, 0);
  return out.str();
}

bool OwareState::IsTerminal() const {
  // Terminate when one player has more than half of the seeds
  // (works both for even and odd number of seeds), or when all seeds
  // are equally shared.
  // A player wins as soon as they have more than half of all seeds. If the
  // seeds are split equally, the game is also over as a draw.
  const int limit = total_seeds_ / 2;
  return board_.score[0] > limit || board_.score[1] > limit ||
         (board_.score[0] == limit && board_.score[1] == limit);
}

std::vector<double> OwareState::Returns() const {
  if (IsTerminal()) {
    if (board_.score[0] > board_.score[1]) {
      return {1, -1};
    } else if (board_.score[0] < board_.score[1]) {
      return {-1, 1};
    } else {
      return {0, 0};
    }
  } else {
    return {0, 0};
  }
}

std::unique_ptr<State> OwareState::Clone() const {
  return std::unique_ptr<State>(new OwareState(*this));
}

int OwareState::DistributeSeeds(int house) {
  // Sowing is deliberately implemented one seed at a time. This makes the
  // wraparound and the rule "skip the source house" explicit.
  int to_distribute = board_.seeds[house];
  SPIEL_CHECK_NE(to_distribute, 0);
  board_.seeds[house] = 0;
  int index = house;
  while (to_distribute > 0) {
    index = (index + 1) % NumHouses();
    // Seeds are never sown into the house they were drawn from, even after
    // making a complete circuit of the board.
    if (index != house) {
      board_.seeds[index]++;
      to_distribute--;
    }
  }
  return index;
}

bool OwareState::InOpponentRow(int house) const {
  // Each row is identified by integer division: indices 0..5 belong to row 0
  // and indices 6..11 belong to row 1 in the default six-house game.
  return (house / num_houses_per_player_) != board_.current_player;
}

bool OwareState::IsGrandSlam(int house) const {
  // Capturing starts at the last house sown into and moves backwards through
  // the opponent's row. A Grand Slam would remove every opponent seed, which
  // Oware forbids; the move remains valid but the capture is cancelled.
  // If there are seeds beyond the house in which the last seed was dropped,
  // it is not a Grand Slam.
  for (int index = UpperHouse(house); index > house; index--) {
    if (board_.seeds[index] > 0) {
      return false;
    }
  }
  // If not all houses are captured starting from the house in which the last
  // seed was dropped, it is not a Grand Slam. It means the opponent will still
  // have some seeds left because none of these houses can be empty due to
  // the way seeds are sown.
  const int lower = LowerHouse(house);
  for (int index = house; index >= lower; index--) {
    SPIEL_CHECK_GT(board_.seeds[index], 0);
    if (!ShouldCapture(board_.seeds[index])) {
      return false;
    }
  }
  return true;
}

int OwareState::OpponentSeeds() const {
  // Count only seeds still in the opponent's row. Captured seeds are already
  // in score[] and do not count as seeds available to receive or capture.
  int count = 0;
  const Player opponent = 1 - board_.current_player;
  const int lower = PlayerLowerHouse(opponent);
  const int upper = PlayerUpperHouse(opponent);
  for (int house = lower; house <= upper; house++) {
    count += board_.seeds[house];
  }
  return count;
}

int OwareState::DoCaptureFrom(int house) {
  // Capture consecutive opponent houses backwards while each contains 2 or
  // 3 seeds. The first house outside that range stops the capture sequence.
  const int lower = LowerHouse(house);
  int captured = 0;
  for (int index = house; index >= lower; index--) {
    if (ShouldCapture(board_.seeds[index])) {
      captured += board_.seeds[index];
      board_.seeds[index] = 0;
    } else {
      break;
    }
  }
  board_.score[board_.current_player] += captured;
  return captured;
}

void OwareState::DoApplyAction(Action action) {
  SPIEL_CHECK_LT(history_.size(), kMaxGameLength);

  // OpenSpiel supplies a row-relative action. Convert it before changing the
  // board, then sow all seeds from that physical house.
  int last_house = DistributeSeeds(ActionToHouse(CurrentPlayer(), action));

  // Captures are possible only when the final seed landed in the opponent's
  // row. Grand Slam moves sow normally but capture nothing.
  if (InOpponentRow(last_house) && !IsGrandSlam(last_house)) {
    const int captured = DoCaptureFrom(last_house);
    if (captured > 0) {
      // No need to keep previous boards for checking game repetition because
      // captured seeds do not re-enter the game.
      boards_since_last_capture_.clear();
    }
  }
  // The turn changes after sowing and any capture has completed.
  board_.current_player = 1 - board_.current_player;

  // insert() returns false when this exact position has already occurred.
  // The current player is included in board_, so turn order matters.
  if (!boards_since_last_capture_.insert(board_).second) {
    // We have game repetition, the game is ended.
    CollectAndTerminate();
  }

  // A player with no legal move cannot continue. Remaining seeds are scored
  // for their respective rows.
  if (LegalActions().empty()) {
    CollectAndTerminate();
  }
}

void OwareState::CollectAndTerminate() {
  // Seeds left on the board are awarded to the player who owns their row.
  // Setting them to zero keeps the board consistent with the scores.
  for (int house = 0; house < NumHouses(); house++) {
    const Player player = house / num_houses_per_player_;
    board_.score[player] += board_.seeds[house];
    board_.seeds[house] = 0;
  }
}

std::string OwareState::ObservationString(Player player) const {
  SPIEL_CHECK_GE(player, 0);
  SPIEL_CHECK_LT(player, num_players_);
  return board_.ToString();
}

void OwareState::ObservationTensor(Player player,
                                   absl::Span<float> values) const {
  SPIEL_CHECK_GE(player, 0);
  SPIEL_CHECK_LT(player, num_players_);

  SPIEL_CHECK_EQ(values.size(), /*seeds*/ NumHouses() + /*scores*/ kNumPlayers);
  // The neural-network input is a normalized flat vector: all house counts
  // first, followed by both players' scores. This implementation exposes the
  // same absolute board to either player because Oware has perfect information.
  for (int house = 0; house < NumHouses(); ++house) {
    values[house] = ((double)board_.seeds[house]) / total_seeds_;
  }
  for (Player player = 0; player < kNumPlayers; ++player) {
    values[NumHouses() + player] =
        ((double)board_.score[player]) / total_seeds_;
  }
}

OwareGame::OwareGame(const GameParameters& params)
    : Game(kGameType, params),
      num_houses_per_player_(ParameterValue<int>("num_houses_per_player")),
      num_seeds_per_house_(ParameterValue<int>("num_seeds_per_house")) {}

std::vector<int> OwareGame::ObservationTensorShape() const {
  return {/*seeds*/ num_houses_per_player_ * kNumPlayers +
          /*scores*/ kNumPlayers};
}

}  // namespace oware
}  // namespace open_spiel

File: Model/oware/oware.py
"""Python implementation of the OpenSpiel Oware game."""

import numpy as np
import pyspiel


_NUM_PLAYERS = 2
_DEFAULT_HOUSES_PER_PLAYER = 6
_DEFAULT_SEEDS_PER_HOUSE = 4
_MIN_CAPTURE = 2
_MAX_CAPTURE = 3
_MAX_GAME_LENGTH = 1000


_GAME_TYPE = pyspiel.GameType(
    short_name="oware",
    long_name="Oware",
    dynamics=pyspiel.GameType.Dynamics.SEQUENTIAL,
    chance_mode=pyspiel.GameType.ChanceMode.DETERMINISTIC,
    information=pyspiel.GameType.Information.PERFECT_INFORMATION,
    utility=pyspiel.GameType.Utility.ZERO_SUM,
    reward_model=pyspiel.GameType.RewardModel.TERMINAL,
    max_num_players=_NUM_PLAYERS,
    min_num_players=_NUM_PLAYERS,
    # The C++ Oware implementation exposes observations, not separate
    # information-state encodings. The observer below provides the RL input.
    provides_information_state_string=False,
    provides_information_state_tensor=False,
    provides_observation_string=True,
    provides_observation_tensor=True,
    parameter_specification={
        "num_houses_per_player": _DEFAULT_HOUSES_PER_PLAYER,
        "num_seeds_per_house": _DEFAULT_SEEDS_PER_HOUSE,
    },
)


_GAME_INFO = pyspiel.GameInfo(
    num_distinct_actions=_DEFAULT_HOUSES_PER_PLAYER,
    max_chance_outcomes=0,
    num_players=_NUM_PLAYERS,
    min_utility=-1.0,
    max_utility=1.0,
    utility_sum=0.0,
    max_game_length=_MAX_GAME_LENGTH,
)


class AyoGame(pyspiel.Game):
    """Game-level configuration shared by all Ayo/Oware states."""

    def __init__(self, params=None):
        """Create a game using the supplied board-size parameters.

        Returns: None.
        """
        # Example input: {"num_houses_per_player": 6, "num_seeds_per_house": 4}
        params = params or {}
        super().__init__(_GAME_TYPE, _GAME_INFO, params)
        self.num_houses_per_player = int(
            params.get("num_houses_per_player", _DEFAULT_HOUSES_PER_PLAYER)
        )
        self.num_seeds_per_house = int(
            params.get("num_seeds_per_house", _DEFAULT_SEEDS_PER_HOUSE)
        )

    def new_initial_state(self):
        """Return a new game state in the starting position.

        Returns: A new AyoState object.
        """
        # Example input: no arguments (call as game.new_initial_state())
        return AyoState(self)

    def make_py_observer(self, iig_obs_type=None, params=None):
        """Create the observer used to convert states into RL inputs.

        Returns: A new AyoObserver object.
        """
        # Example input: iig_obs_type=None, params=None
        return AyoObserver(
            iig_obs_type or pyspiel.IIGObservationType(perfect_recall=False),
            params or {},
            self,
        )


class AyoState(pyspiel.State):
    """One mutable position in an Oware game.

    ``board`` is laid out in sowing order:

      player 0: board[0:6]
      player 1: board[6:12]

    Actions are local to the current player's row. Thus action 0 means
    ``board[0]`` for player 0 and ``board[6]`` for player 1.
    """

    def __init__(self, game):
        """Initialize the board, scores, turn, and repetition tracking.

        Returns: None.
        """
        # Example input: game = AyoGame()
        super().__init__(game)
        self.num_houses_per_player = game.num_houses_per_player
        self.num_houses = _NUM_PLAYERS * self.num_houses_per_player
        self.total_seeds = (
            _NUM_PLAYERS
            * self.num_houses_per_player
            * game.num_seeds_per_house
        )
        self.board = [game.num_seeds_per_house] * self.num_houses
        self.captured = [0, 0]
        self._current_player = 0
        self._game_over = False
        self._returns = [0.0, 0.0]
        self._positions_since_capture = {self._position_key()}

    def _position_key(self):
        """Return a hashable representation of the current game position.

        Returns: A tuple containing the player, scores, and board.
        """
        # Example input: no arguments (call as state._position_key())
        return (self._current_player, tuple(self.captured), tuple(self.board))

    def current_player(self):
        """Return the current player, or TERMINAL after the game ends.

        Returns: The current player number or pyspiel.PlayerId.TERMINAL.
        """
        # Example input: no arguments (call as state.current_player())
        return pyspiel.PlayerId.TERMINAL if self._game_over else self._current_player

    def _player_lower_house(self, player):
        """Return the first board index belonging to a player.

        Returns: An integer board index.
        """
        # Example input: player=1
        return player * self.num_houses_per_player

    def _player_upper_house(self, player):
        """Return the last board index belonging to a player.

        Returns: An integer board index.
        """
        # Example input: player=1
        return self._player_lower_house(player) + self.num_houses_per_player - 1

    def _lower_house(self, house):
        """Return the first index in the row containing a house.

        Returns: An integer board index.
        """
        # Example input: house=8
        return (house // self.num_houses_per_player) * self.num_houses_per_player

    def _upper_house(self, house):
        """Return the last index in the row containing a house.

        Returns: An integer board index.
        """
        # Example input: house=8
        return self._lower_house(house) + self.num_houses_per_player - 1

    def _action_to_house(self, player, action):
        """Convert a player's local action into a board index.

        Returns: An integer board index.
        """
        # Example input: player=1, action=2
        return player * self.num_houses_per_player + action

    def _house_to_action(self, house):
        """Convert a board index into its local action number.

        Returns: An integer action number.
        """
        # Example input: house=8
        return house % self.num_houses_per_player

    def _opponent_seeds(self):
        """Return the number of seeds currently in the opponent's row.

        Returns: The opponent's seed count as an integer.
        """
        # Example input: no arguments (call as state._opponent_seeds())
        opponent = 1 - self._current_player
        lower = self._player_lower_house(opponent)
        upper = self._player_upper_house(opponent)
        return sum(self.board[lower : upper + 1])

    def _legal_actions(self, player):
        """Return the actions the specified player may currently choose.

        Returns: A list of legal action numbers.
        """
        # Example input: player=0
        if self._game_over or player not in (0, 1):
            return []

        lower = self._player_lower_house(self._current_player)
        upper = self._player_upper_house(self._current_player)

        if self._opponent_seeds() == 0:
            # The move must reach the opponent row and give it a seed.
            return [
                self._house_to_action(house)
                for house in range(lower, upper + 1)
                if self.board[house] - (upper - house) > 0
            ]

        return [
            self._house_to_action(house)
            for house in range(lower, upper + 1)
            if self.board[house] > 0
        ]

    def _distribute_seeds(self, house):
        """Sow one seed at a time, skipping the source house.

        Returns: The index of the house receiving the final seed.
        """
        # Example input: house=2
        to_distribute = self.board[house]
        if to_distribute == 0:
            raise ValueError("Cannot sow from an empty house")
        self.board[house] = 0
        index = house
        while to_distribute > 0:
            index = (index + 1) % self.num_houses
            if index != house:
                self.board[index] += 1
                to_distribute -= 1
        return index

    def _in_opponent_row(self, house):
        """Return whether a house belongs to the opponent's row.

        Returns: True if the house is in the opponent's row; otherwise False.
        """
        # Example input: house=8
        return house // self.num_houses_per_player != self._current_player

    @staticmethod
    def _should_capture(seeds):
        """Return whether a house with this seed count can be captured.

        Returns: True if the seed count is capturable; otherwise False.
        """
        # Example input: seeds=2
        # Capture rules in this variant arent the same as oware 
        return _MIN_CAPTURE <= seeds <= _MAX_CAPTURE

    def _is_grand_slam(self, house):
        """Return whether capturing from ``house`` empties the opponent row.

        Returns: True if the capture empties the opponent's row; otherwise False.
        """
        # Example input: house=8
        for index in range(self._upper_house(house), house, -1):
            if self.board[index] > 0:
                return False
        lower = self._lower_house(house)

        # For the variant i am considering the grand slam rule is applied but backwards capture rule isnt applied 
        return all(
            self.board[index] > 0
            and self._should_capture(self.board[index])
            for index in range(house, lower - 1, -1)
        )

    def _capture_from(self, house):
        """Capture consecutive 2- or 3-seed opponent houses backwards.

        Returns: The number of seeds captured.
        """
        # Example input: house=8
        #Also this capture function will be modified for the ayo olopon variant 
        captured = 0
        lower = self._lower_house(house)
        for index in range(house, lower - 1, -1):
            if not self._should_capture(self.board[index]):
                break
            captured += self.board[index]
            self.board[index] = 0
        self.captured[self._current_player] += captured
        return captured

    def _score_terminal(self):
        """Return whether the captured-seed scores meet a terminal condition.

        Returns: True if a score-based terminal condition is met; otherwise False.
        """
        # Example input: no arguments (call as state._score_terminal())
        limit = self.total_seeds // 2
        return (
            self.captured[0] > limit
            or self.captured[1] > limit
            or (self.captured[0] == limit and self.captured[1] == limit)
        )

    def _set_score_returns_and_end(self):
        """Mark the game finished and assign returns from the final scores.

        Returns: None.
        """
        # Example input: no arguments (call as state._set_score_returns_and_end())
        self._game_over = True
        if self.captured[0] > self.captured[1]:
            self._returns = [1.0, -1.0]
        elif self.captured[0] < self.captured[1]:
            self._returns = [-1.0, 1.0]
        else:
            self._returns = [0.0, 0.0]

    def _collect_and_terminate(self):
        """Award remaining seeds to the owners of their rows.

        Returns: None.
        """
        # Example input: no arguments (call as state._collect_and_terminate())
        for house, seeds in enumerate(self.board):
            owner = house // self.num_houses_per_player
            self.captured[owner] += seeds
            self.board[house] = 0
        self._set_score_returns_and_end()

    def _apply_action(self, action):
        """Apply an action, including sowing, captures, and end checks.

        Returns: None.
        """
        # Example input: action=2
        if action not in self._legal_actions(self._current_player):
            raise ValueError(f"Illegal action: {action}")

        house = self._action_to_house(self._current_player, action)
        last_house = self._distribute_seeds(house)

        if self._in_opponent_row(last_house) and not self._is_grand_slam(
            last_house
        ):
            if self._capture_from(last_house) > 0:
                # Captured seeds cannot return to the board, so earlier
                # positions cannot recur after a capture.
                self._positions_since_capture.clear()

        self._current_player = 1 - self._current_player

        if self._score_terminal():
            self._set_score_returns_and_end()
            return

        if self._position_key() in self._positions_since_capture:
            self._collect_and_terminate()
            return
        self._positions_since_capture.add(self._position_key())

        if not self._legal_actions(self._current_player):
            self._collect_and_terminate()

    def is_terminal(self):
        """Return whether the game has reached a terminal state.

        Returns: True if the game is over; otherwise False.
        """
        # Example input: no arguments (call as state.is_terminal())
        return self._game_over

    def returns(self):
        """Return final payoffs, or zero payoffs while play continues.

        Returns: A list containing the two players' payoffs.
        """
        # Example input: no arguments (call as state.returns())
        return list(self._returns) if self._game_over else [0.0, 0.0]

    def _action_to_string(self, player, action):
        """Return the display label for a player's action.

        Returns: A string action label such as ``A2`` or ``a2``.
        """
        # Example input: player=0, action=2
        return f"{'A' if player == 0 else 'a'}{action}"

    def __str__(self):
        """Return a readable summary of the board and game status.

        Returns: A string containing the board, scores, player, and status.
        """
        # Example input: no arguments (call as str(state))
        return (
            f"Board: {self.board}\n"
            f"Captured: {self.captured}\n"
            f"Current player: {self.current_player()}\n"
            f"Terminal: {self._game_over}"
        )


class AyoObserver:
    """Flat normalized observation: houses followed by both scores."""

    def __init__(self, iig_obs_type, params, game):
        """Create storage for a normalized observation tensor.

        Returns: None.
        """
        # Example input: iig_obs_type=None, params={}, game=AyoGame()
        del iig_obs_type
        if params:
            raise ValueError(f"Observation parameters are not supported: {params}")
        size = 2 * game.num_houses_per_player + _NUM_PLAYERS
        self.tensor = np.zeros(size, dtype=np.float32)
        self.dict = {"observation": self.tensor}

    def set_from(self, state, player):
        """Copy a game state into the observer's normalized tensor.

        Returns: None.
        """
        # Example input: state=game.new_initial_state(), player=0
        del player
        self.tensor[: state.num_houses] = np.asarray(state.board) / state.total_seeds
        self.tensor[state.num_houses :] = np.asarray(state.captured) / state.total_seeds

    def string_from(self, state, player):
        """Return a string representation of the observed state.

        Returns: A string representation of the state.
        """
        # Example input: state=game.new_initial_state(), player=0
        del player
        return str(state)


pyspiel.register_game(_GAME_TYPE, AyoGame)
